Apply default values during property expansion

**Description**
Default values are not applied during expansion of config values. E.g. given a config in _application.properties_ like this:

    quarkus.kafka-streams.application-server=${hostname}:${quarkus.http.port}

Then the HTTP port should default to the value from the config item, but instead an error is raised:

    Build step io.quarkus.deployment.steps.ConfigurationSetup#initializeConfiguration threw an exception: java.util.NoSuchElementException: Property quarkus.http.port not found

Things work when specifying a value explicitly for `quarkus.http.port` or when giving a default value within the expanded expression (${quarkus.http.port:8080}).

CC @dmlloyd

@dmlloyd, here's the complete stracktrace:

```
org.junit.jupiter.api.extension.TestInstantiationException:
TestInstanceFactory [io.quarkus.test.junit.QuarkusTestExtension] failed to instantiate test class [io.quarkus.it.kafka.streams.KafkaStreamsTest]: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.ConfigurationSetup#initializeConfiguration threw an exception: java.util.NoSuchElementException: Property quarkus.http.port not found
	[error]: Build step io.quarkus.kafka.streams.deployment.KafkaStreamsProcessor#processBuildTimeConfig threw an exception: java.util.NoSuchElementException: Property quarkus.http.port not found
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.invokeTestInstanceFactory(ClassTestDescriptor.java:314)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:289)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateTestClass(ClassTestDescriptor.java:281)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.instantiateAndPostProcessTestInstance(ClassTestDescriptor.java:269)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.lambda$testInstancesProvider$2(ClassTestDescriptor.java:259)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.lambda$testInstancesProvider$3(ClassTestDescriptor.java:263)
	at java.util.Optional.orElseGet(Optional.java:267)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.lambda$testInstancesProvider$4(ClassTestDescriptor.java:262)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$prepare$0(TestMethodTestDescriptor.java:98)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:97)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.prepare(TestMethodTestDescriptor.java:68)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$prepare$1(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.prepare(NodeTestTask.java:107)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:75)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:170)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:154)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:90)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:142)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:117)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:383)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:344)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:125)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:417)
Caused by: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.ConfigurationSetup#initializeConfiguration threw an exception: java.util.NoSuchElementException: Property quarkus.http.port not found
	[error]: Build step io.quarkus.kafka.streams.deployment.KafkaStreamsProcessor#processBuildTimeConfig threw an exception: java.util.NoSuchElementException: Property quarkus.http.port not found
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:129)
	at io.quarkus.test.junit.QuarkusTestExtension.doJavaStart(QuarkusTestExtension.java:220)
	at io.quarkus.test.junit.QuarkusTestExtension.createTestInstance(QuarkusTestExtension.java:286)
	at org.junit.jupiter.engine.descriptor.ClassTestDescriptor.invokeTestInstanceFactory(ClassTestDescriptor.java:299)
	... 47 more
Caused by: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.deployment.steps.ConfigurationSetup#initializeConfiguration threw an exception: java.util.NoSuchElementException: Property quarkus.http.port not found
	[error]: Build step io.quarkus.kafka.streams.deployment.KafkaStreamsProcessor#processBuildTimeConfig threw an exception: java.util.NoSuchElementException: Property quarkus.http.port not found
	at io.quarkus.builder.Execution.run(Execution.java:108)
	at io.quarkus.builder.BuildExecutionBuilder.execute(BuildExecutionBuilder.java:121)
	at io.quarkus.deployment.QuarkusAugmentor.run(QuarkusAugmentor.java:97)
	at io.quarkus.runner.RuntimeRunner.run(RuntimeRunner.java:94)
	... 50 more
	Suppressed: java.util.NoSuchElementException: Property quarkus.http.port not found
		at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:89)
		at io.quarkus.runtime.configuration.ConfigExpander.accept(ConfigExpander.java:55)
		at io.quarkus.runtime.configuration.ConfigExpander.accept(ConfigExpander.java:15)
		at org.wildfly.common.expression.ExpressionNode.emit(ExpressionNode.java:42)
		at org.wildfly.common.expression.CompositeNode.emit(CompositeNode.java:42)
		at org.wildfly.common.expression.Expression.evaluateException(Expression.java:75)
		at org.wildfly.common.expression.Expression.evaluate(Expression.java:89)
		at io.quarkus.runtime.configuration.ExpandingConfigSource.expandValue(ExpandingConfigSource.java:76)
		at io.quarkus.runtime.configuration.ExpandingConfigSource.expand(ExpandingConfigSource.java:48)
		at io.quarkus.runtime.configuration.ExpandingConfigSource.getValue(ExpandingConfigSource.java:44)
		at io.quarkus.runtime.configuration.DeploymentProfileConfigSource.getValue(DeploymentProfileConfigSource.java:57)
		at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:74)
		at io.quarkus.kafka.streams.deployment.KafkaStreamsProcessor.processBuildTimeConfig(KafkaStreamsProcessor.java:70)
		at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
		at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
		at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
		at java.lang.reflect.Method.invoke(Method.java:498)
		at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:507)
		at io.quarkus.builder.BuildContext.run(BuildContext.java:398)
		at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
		at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
		at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1538)
		at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1429)
		at java.lang.Thread.run(Thread.java:748)
		at org.jboss.threads.JBossThread.run(JBossThread.java:479)
Caused by: java.util.NoSuchElementException: Property quarkus.http.port not found
	at io.smallrye.config.SmallRyeConfig.getValue(SmallRyeConfig.java:89)
	at io.quarkus.runtime.configuration.ConfigExpander.accept(ConfigExpander.java:55)
	at io.quarkus.runtime.configuration.ConfigExpander.accept(ConfigExpander.java:15)
	at org.wildfly.common.expression.ExpressionNode.emit(ExpressionNode.java:42)
	at org.wildfly.common.expression.CompositeNode.emit(CompositeNode.java:42)
	at org.wildfly.common.expression.Expression.evaluateException(Expression.java:75)
	at org.wildfly.common.expression.Expression.evaluate(Expression.java:89)
	at io.quarkus.runtime.configuration.ExpandingConfigSource.expandValue(ExpandingConfigSource.java:76)
	at io.quarkus.runtime.configuration.ExpandingConfigSource.expand(ExpandingConfigSource.java:48)
	at io.quarkus.runtime.configuration.ExpandingConfigSource.getValue(ExpandingConfigSource.java:44)
	at io.quarkus.runtime.configuration.DeploymentProfileConfigSource.getValue(DeploymentProfileConfigSource.java:57)
	at io.smallrye.config.SmallRyeConfig.getOptionalValue(SmallRyeConfig.java:95)
	at io.quarkus.deployment.configuration.OptionalObjectConfigType.acceptConfigurationValueIntoGroup(OptionalObjectConfigType.java:78)
	at io.quarkus.deployment.configuration.GroupConfigType.acceptConfigurationValueIntoLeaf(GroupConfigType.java:238)
	at io.quarkus.deployment.configuration.OptionalObjectConfigType.acceptConfigurationValue(OptionalObjectConfigType.java:33)
	at io.quarkus.deployment.configuration.ConfigDefinition.loadConfiguration(ConfigDefinition.java:389)
	at io.quarkus.deployment.steps.ConfigurationSetup.initializeConfiguration(ConfigurationSetup.java:239)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at io.quarkus.deployment.ExtensionLoader$1.execute(ExtensionLoader.java:507)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:398)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:2011)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1538)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1429)
	at java.lang.Thread.run(Thread.java:748)
	at org.jboss.threads.JBossThread.run(JBossThread.java:479)
```

> Also where is that property quarkus.kafka-streams.application-server defined? Is that a pass-through property, and if so, what's facilitating the pass-through?

It's a regular property, [defined](https://github.com/quarkusio/quarkus/blob/c7b990b0377a25cf9bcc09de9876c616fcd6f454/extensions/kafka-streams/runtime/src/main/java/io/quarkus/kafka/streams/runtime/KafkaStreamsRuntimeConfig.java#L28) as a config item.

--- a/core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigDefinition.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/configuration/ConfigDefinition.java
@@ -73,11 +73,17 @@ public class ConfigDefinition extends CompoundConfigType {
     private final TreeMap<String, RootInfo> rootTypesByContainingName = new TreeMap<>();
     private final FieldDescriptor rootField;
     private final TreeMap<String, String> loadedProperties = new TreeMap<>();
+    private final boolean deferResolution;
 
-    public ConfigDefinition(final FieldDescriptor rootField) {
+    public ConfigDefinition(final FieldDescriptor rootField, final boolean deferResolution) {
         super(null, null, false);
+        this.deferResolution = deferResolution;
         Assert.checkNotNullParam("rootField", rootField);
         this.rootField = rootField;
+    }
+
+    public ConfigDefinition(final FieldDescriptor rootField) {
+        this(rootField, false);
     }
 
     void acceptConfigurationValueIntoLeaf(final LeafConfigType leafType, final NameIterator name,
@@ -388,7 +394,16 @@ public class ConfigDefinition extends CompoundConfigType {
                             name.goToEnd();
                             leafType.acceptConfigurationValue(name, cache, config);
                             final String nameString = name.toString();
-                            definition.loadedProperties.put(nameString, config.getValue(nameString, String.class));
+                            if (definition.deferResolution) {
+                                boolean old = ExpandingConfigSource.setExpanding(false);
+                                try {
+                                    definition.loadedProperties.put(nameString, config.getValue(nameString, String.class));
+                                } finally {
+                                    ExpandingConfigSource.setExpanding(old);
+                                }
+                            } else {
+                                definition.loadedProperties.put(nameString, config.getValue(nameString, String.class));
+                            }
                             continue outer;
                         }
                     }

--- a/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/steps/ConfigurationSetup.java
@@ -189,7 +189,7 @@ public class ConfigurationSetup {
         // set up the configuration definitions
         final ConfigDefinition buildTimeConfig = new ConfigDefinition(FieldDescriptor.of("Bogus", "No field", "Nothing"));
         final ConfigDefinition buildTimeRunTimeConfig = new ConfigDefinition(BUILD_TIME_CONFIG_FIELD);
-        final ConfigDefinition runTimeConfig = new ConfigDefinition(RUN_TIME_CONFIG_FIELD);
+        final ConfigDefinition runTimeConfig = new ConfigDefinition(RUN_TIME_CONFIG_FIELD, true);
 
         // populate it with all known types
         for (Class<?> clazz : ServiceUtil.classesNamedIn(extensionClassLoaderBuildItem.getExtensionClassLoader(),

